### PR TITLE
CORS-4537: Add API-backed disk type availability validation

### PR DESCRIPTION
**Pull request by barbacbd**
Disk type validation was previously entirely static, relying on
    hardcoded maps (InstanceTypeToDiskTypeMap, ComputeSupportedDisks) that
    encode which disk types are compatible with which instance families.
    This works for public GCP where all standard disk types are available
    in every region, but fails for sovereign clouds where regional
    availability differs. For example, pd-ssd is not available in
    u-germany-northeast1, but the static map would allow c3 + pd-ssd
    since c3 supports pd-ssd in public GCP.
    
    This change adds an API-backed availability check that queries each
    zone individually via DiskTypes.Get, following the existing pattern
    established by GetMachineTypeWithZones for machine type validation.
    Per-zone Get calls are used instead of DiskTypes.AggregatedList because
    the aggregated list does not return results on sovereign clouds.
    
    The two validation layers are complementary:
    
    1. Static check (validateInstanceAndDiskType): Is this disk type
       compatible with this instance family? Uses the hardcoded
       InstanceTypeToDiskTypeMap. Catches combinations that never work
       anywhere (e.g. n2 + hyperdisk-balanced).
    
    2. API check (validateDiskTypeAvailability): Is this disk type
       available in this specific region and zones? Calls the GCP API.
       Catches disk types that are valid in general but unavailable in the
       target region (e.g. pd-ssd in sovereign cloud regions).
    
    The API check runs after the static check passes, so incompatible
    combinations are caught early without an API call. If the DiskTypes
    API is unreachable due to a non-googleapi error (e.g. network issues,
    missing permissions), validation degrades gracefully by logging a
    warning and continuing with only the static check.

## Summary by CodeRabbit

* **New Features**
  * GCP install configuration now validates disk type availability across regions and selected zones.
  * Added support for detecting disk types and their available zones.
  * Improved handling of sovereign GCP environments and custom universe domains.

* **Bug Fixes**
  * Improved error handling for GCP resource and validation failures.
  * Corrected disk type defaults and availability checks for sovereign projects.

* **Tests**
  * Added coverage for disk type validation, sovereign environments, universe domains, and disk type defaults.

### CODE REVIEW

File: pkg/asset/installconfig/gcp/validation.go [hunk 1]
```diff
@@ -203,6 +237,21 @@ func validateServiceAccountPresent(client API, ic *types.InstallConfig) field.Er
...
+
+// DefaultInstanceTypeForArchAndProjectID returns the appropriate instance type based on the target architecture and project ID.
+// For sovereign cloud environments, it returns c3-standard-4 which is available in those regions.
+// For public GCP, it returns n2-standard-4 (x86) or t2a-standard-4 (ARM64).
+func DefaultInstanceTypeForArchAndProjectID(arch types.Architecture, projectID string) string {
+	cloudEnv := gcp.GetCloudEnvironment(projectID)
+
+	// Sovereign cloud uses c3-standard-4 for all architectures
+	if cloudEnv == gcp.CloudEnvironmentSovereign {
+		return "c3-standard-4"
```
[the remark this answers is not in the document]
Comment (barbacbd) [reply pkg/asset/installconfig/gcp/validation.go]: Disregarding this for now. Sovereign clouds only have A3, C3, and M3 instance families available — none of which support ARM64. The current behavior returns `c3-standard-4` as the sovereign default, and the existing ARM64 validation at line 95 will reject it with a clear error about unsupported families. This is the correct behavior since ARM64 is not supported on sovereign clouds.

File: pkg/asset/installconfig/gcp/client.go [hunk 2]
```diff
@@ -223,6 +224,49 @@ func (c *Client) GetMachineTypeWithZones(ctx context.Context, project, region, m
...
+			found = dt
+		}
+		zones.Insert(zone.Name)
+	}
+
+	if found == nil {
+		return nil, nil, nil
+	}
+
+	return found, zones, nil
```
[the remark this answers is not in the document]
Comment (barbacbd) [reply pkg/asset/installconfig/gcp/client.go:268]: The validation tests in `validation_test.go` already cover the behavior of `GetDiskTypeWithZones` at the API interface level via gomock — including disk type found, not found (nil return), zone filtering, and error propagation scenarios in `TestValidateDiskTypeAvailability`. Testing the concrete `Client` implementation directly would require mocking the GCP compute HTTP transport, which is heavy lift for limited additional coverage beyond what the interface-level tests already provide.

File: pkg/types/gcp/platform.go [hunk 3]
```diff
@@ -193,3 +208,28 @@ func GetDefaultServiceAccount(platform *Platform, clusterID string, role string)
...
+	// No known sovereign prefix found
+	return ""
+}
+
+// IsNonDefaultUniverseDomain returns true if the universe domain is non-empty
+// and not the default "googleapis.com" value.
+// This indicates a sovereign cloud or custom universe domain configuration.
+func IsNonDefaultUniverseDomain(universeDomain string) bool {
+	return universeDomain != "" && universeDomain != "googleapis.com"
+}
```
[the remark this answers is not in the document]
Comment (barbacbd) [reply pkg/types/gcp/platform.go]: Added unit tests for the sovereign-aware helpers:
- `TestGetCloudEnvironment` — covers sovereign prefix detection, public GCP projects, org-scoped projects, and edge cases
- `TestIsNonDefaultUniverseDomain` — covers default domain, empty string, and custom domains
- `TestDefaultDiskTypeForInstanceAndProjectID` — covers public vs sovereign disk type selection, fallback behavior, and unrecognized instance types

File: pkg/clusterapi/system.go [hunk 4]
```diff
@@ -307,6 +307,18 @@ func (c *system) Run(ctx context.Context) error { //nolint:gocyclo
...
+		// set for the CAPI provider to use the correct GCP API endpoints
+		universeDomain, err := session.Credentials.GetUniverseDomain()
+		if err != nil {
+			return fmt.Errorf("failed to get universe domain from GCP credentials: %w", err)
+		}
+		if gcp.IsNonDefaultUniverseDomain(universeDomain) {
+			capgEnvVars["GOOGLE_CLOUD_UNIVERSE_DOMAIN"] = universeDomain
+			logrus.Infof("setting GOOGLE_CLOUD_UNIVERSE_DOMAIN to %s for capg infrastructure controller", universeDomain)
+		}
+
```
[the remark this answers is not in the document]
Comment (barbacbd) [reply pkg/clusterapi/system.go]: Good catch — removed the duplicate block. The `gcp.IsNonDefaultUniverseDomain` check is now the single source of truth.

File: pkg/asset/installconfig/gcp/session.go [hunk 5]
```diff
@@ -196,7 +196,10 @@ func (f *contentLoader) Content() string {
 type cliLoader struct{}
 
 func (c *cliLoader) Load(ctx context.Context) (*googleoauth.Credentials, error) {
-	return googleoauth.FindDefaultCredentials(ctx, compute.CloudPlatformScope)
+	// Use FindDefaultCredentialsWithParams to ensure universe domain from credentials is applied
+	return googleoauth.FindDefaultCredentialsWithParams(ctx, googleoauth.CredentialsParams{
```
Comment (rna-afk) [inline pkg/asset/installconfig/gcp/session.go]: Is there a difference between the code? Looks like `FindDefaultCredentials` does the exact same thing but I'm guessing there's a practical difference.

File: pkg/asset/machines/worker.go [hunk 6]
```diff
@@ -141,13 +141,13 @@ func defaultAzureMachinePoolPlatform(env azuretypes.CloudEnvironment) azuretypes
 	}
 }
 
-func defaultGCPMachinePoolPlatform(arch types.Architecture) gcptypes.MachinePool {
-	instanceType := icgcp.DefaultInstanceTypeForArch(arch)
+func defaultGCPMachinePoolPlatform(arch types.Architecture, projectID, region string) gcptypes.MachinePool {
```
Comment (tthvo) [inline pkg/asset/machines/worker.go]:
```suggestion
func defaultGCPMachinePoolPlatform(arch types.Architecture, projectID) gcptypes.MachinePool {
```

We can remove `region` as it is unused, right?

File: pkg/clusterapi/system.go [hunk 7]
```diff
@@ -308,19 +307,20 @@ func (c *system) Run(ctx context.Context) error { //nolint:gocyclo
...
 		if err != nil {
-			return fmt.Errorf("failed to get universe domain from gcp credentials: %w", err)
+			return fmt.Errorf("failed to get universe domain from GCP credentials: %w", err)
 		}
-		if ud != "googleapis.com" {
-			capgEnvVars["GOOGLE_CLOUD_UNIVERSE_DOMAIN"] = ud
-			logrus.Infof("setting GOOGLE_CLOUD_UNIVERSE_DOMAIN to %q for capg infrastructure controller", ud)
+		if gcp.IsNonDefaultUniverseDomain(universeDomain) {
+			capgEnvVars["GOOGLE_CLOUD_UNIVERSE_DOMAIN"] = universeDomain
+			logrus.Infof("setting GOOGLE_CLOUD_UNIVERSE_DOMAIN to %s for capg infrastructure controller", universeDomain)
```
Comment (tthvo) [inline pkg/clusterapi/system.go]: Did you mean to include this change or it should be removed to limit the PR scope to disk validations?

File: pkg/types/gcp/platform.go [hunk 8]
```diff
@@ -202,3 +216,28 @@ func GetDefaultServiceAccount(platform *Platform, clusterID string, role string)
...
+// public GCP projects (orgname:project-id).
+func GetCloudEnvironment(projectID string) string {
+	// Check if project ID has a known sovereign cloud prefix
+	if strings.Contains(projectID, ":") {
+		parts := strings.SplitN(projectID, ":", 2)
+		if len(parts) == 2 && sets.New(sovereignCloudProjectPrefixes...).Has(parts[0]) {
+			// Known sovereign prefix is definitive - this IS a sovereign cloud project
+			return CloudEnvironmentSovereign
+		}
+	}
```
Comment (tthvo) [inline pkg/types/gcp/platform.go]: We can skip the `strings.Contains(projectID, ":")` check because `strings.splitN` will return 1 part without `:`, thus making the `if` on line 228 false (same as `GetDefaultServiceAccount`).

```go
// Domain-scoped project IDs (e.g. "eu0:openshift") use a reversed
// dot-separated format in SA emails: "openshift.eu0".
if parts := strings.SplitN(projectID, ":", 2); len(parts) == 2 && sets.New(sovereignCloudProjectPrefixes...).Has(parts[0]) {
  // Known sovereign prefix is definitive - this IS a sovereign cloud project
  return CloudEnvironmentSovereign
}
```

File: pkg/types/gcp/machinepools.go [hunk 9]
```diff
@@ -372,21 +372,43 @@ func GetGCPInstanceFamily(instanceType string) string {
...
-			defaultDiskType = HyperDiskBalanced
-		default:
-			// this shouldn't happen because all supported instance types
-			// have either pd-ssd or hyperdisk balanced
-			defaultDiskType = diskTypes[0]
+		cloudEnv := GetCloudEnvironment(projectID)
+
+		// Sovereign cloud prefers hyperdisk-balanced
+		if cloudEnv == CloudEnvironmentSovereign {
+			switch {
```
Comment (tthvo) [inline pkg/types/gcp/machinepools.go]: I notice the two branches differ only in preference order. How about below where we define a precedence slice of disk types?

```go
func DefaultDiskTypeForInstanceAndProjectID(instanceType, projectID string) string {
	diskTypes, ok := GetDiskTypes(instanceType)
	if !ok {
		return PDSSD
	}

	preferred := []string{PDSSD, HyperDiskBalanced}
	if GetCloudEnvironment(projectID) == CloudEnvironmentSovereign {
		preferred = []string{HyperDiskBalanced, PDSSD}
	}

	supportedDiskTypes := sets.New(diskTypes...)
	for _, dt := range preferred {
		if supportedDiskTypes.Has(dt) {
			return dt
		}
	}
	return diskTypes[0]
}
```

File: pkg/asset/installconfig/gcp/validation.go [hunk 10]
```diff
@@ -186,6 +189,39 @@ func ValidateInstanceType(client API, fieldPath *field.Path, project, region str
...
+		return append(allErrs, field.Invalid(fieldPath.Child("diskType"), diskType, errMsg))
+	}
+
+	userZones := sets.New(zones...)
+	if len(userZones) > 0 {
+		if diff := userZones.Difference(dtZones); len(diff) > 0 {
+			errMsg := fmt.Sprintf("disk type %s is not available in zones: %v", diskType, sets.List(diff))
+			allErrs = append(allErrs, field.Invalid(fieldPath.Child("diskType"), diskType, errMsg))
+		}
+	}
```
Comment (tthvo) [inline pkg/asset/installconfig/gcp/validation.go]: When the user doesn't define zones, the installer defaults to use zones where machine type is available, right 👇?

https://github.com/openshift/installer/blob/91e9633eca636b3b36de7fbad555a475f7e95b7d/pkg/asset/installconfig/gcp/validation.go#L161-L168

In that case, the base zone list (if not defined by users) should be zones where machine type is available. For example, we can:

```golang
// Move validateDiskTypeAvailability call after line 171
// Variable userZones is either user-defined or default zones
// where machine type is available so we should pass that in
// instead of variable zones.
allErrs = append(allErrs, validateDiskTypeAvailability(client,
  fieldPath, project, region,
  sets.List(userZones), // We use the userZones variable instead of zones
  diskType)...)
```

File: pkg/asset/installconfig/gcp/validation.go [hunk 11]
```diff
@@ -186,6 +189,39 @@ func ValidateInstanceType(client API, fieldPath *field.Path, project, region str
...
+
+	dt, dtZones, err := client.GetDiskTypeWithZones(context.TODO(), project, region, diskType)
+	if err != nil {
+		var gerr *googleapi.Error
+		if errors.As(err, &gerr) {
+			return append(allErrs, field.Invalid(fieldPath.Child("diskType"), diskType, err.Error()))
+		}
+		logrus.Warnf("Unable to validate disk type %s availability: %v", diskType, err)
+		return allErrs
```
Comment (tthvo) [inline pkg/asset/installconfig/gcp/validation.go]: I think non-API errors (e.g. DNS failures, TCP timeouts, TLS handshake errors, etc) should fail the validation right? That is: `field.InternalError`.

We should fail fast so bad combination won't come up late during machine provisioning 🤔

File: pkg/asset/installconfig/gcp/validation.go [hunk 12]
```diff
@@ -786,23 +837,50 @@ func validateMarketplaceImages(client API, ic *types.InstallConfig) field.ErrorL
...
+					field.NewPath("controlPlane", "platform", "gcp", "osImage", "project"),
+					"must specify image project for sovereign cloud"))
 			}
+			if osImage.Project != "" {
+				image, err = client.GetImage(context.TODO(), osImage.Name, osImage.Project)
+				if err != nil {
+					allErrs = append(allErrs, field.Invalid(field.NewPath("controlPlane", "platform", "gcp", "osImage"), *osImage, fmt.Sprintf(errorMessage, err)))
+				}
+			}
+		} else if isSovereignCloud && defaultOsImage == nil {
```
Comment (tthvo) [inline pkg/asset/installconfig/gcp/validation.go]: I guess we can drop the validations that requires BYO `osImage` in GCD since we have https://github.com/openshift/installer/pull/10686, right?

File: pkg/asset/installconfig/gcp/validation_test.go [hunk 13]
```diff
@@ -1381,6 +1389,84 @@ func TestValidateInstanceType(t *testing.T) {
...
+			name:           "Disk type not available in region",
+			diskType:       "hyperdisk-balanced",
+			zones:          []string{},
+			mockDiskType:   nil,
+			mockZones:      nil,
+			expectedError:  true,
+			expectedErrMsg: `disk type hyperdisk-balanced is not available in region us-east1`,
+		},
+		{
+			name:           "Disk type not available in specific zones",
```
Comment (tthvo) [inline pkg/asset/installconfig/gcp/validation_test.go:1426]: 💡 We should a case: the disk type is available in fewer zones than the machine type.

File: pkg/asset/installconfig/gcp/validation.go (hunk 12 above)
Comment (tthvo) [reply pkg/asset/installconfig/gcp/validation.go]: If we plan to keep only `if isSovereignCloud && osImage.Project == ""`. Then, we should only validate it for `controlPlane` and `compute` pool since users can populate directly.

Also, `defaultMachinePlatform` is already merged into each `controlPlane` and `compute` pool before validation runs. So, by the time `validateInstanceTypes` runs, each `controlPlane` and `compute` pool already has the `defaultMachinePlatform` values baked in.

https://github.com/openshift/installer/blob/91e9633eca636b3b36de7fbad555a475f7e95b7d/pkg/types/gcp/defaults/machinepool.go#L9-L14

File: pkg/asset/installconfig/gcp/validation.go [hunk 14]
```diff
@@ -186,6 +190,37 @@ func ValidateInstanceType(client API, fieldPath *field.Path, project, region str
...
+		return allErrs
+	}
+
+	dt, dtZones, err := client.GetDiskTypeWithZones(context.TODO(), project, region, diskType)
+	if err != nil {
+		var gerr *googleapi.Error
+		if errors.As(err, &gerr) {
+			return append(allErrs, field.Invalid(fieldPath.Child("diskType"), diskType, err.Error()))
+		}
+		return append(allErrs, field.InternalError(fieldPath.Child("diskType"), err))
```
[the remark this answers is not in the document]
Comment (tthvo) [reply pkg/asset/installconfig/gcp/validation.go:206]: Hmm no, it's better to surface the error than silently ignoring an invalid combination.

File: pkg/clusterapi/system.go [hunk 15]
```diff
@@ -312,17 +312,6 @@ func (c *system) Run(ctx context.Context) error { //nolint:gocyclo
 			logrus.Infof("setting %q to %s for capg infrastructure controller", gAppCredEnvVar, v)
 		}
 
-		// Google Cloud Dedicated support: detect universe domain from
-		// credentials and pass it to the CAPG controller via env var.
-		ud, err := session.Credentials.GetUniverseDomain()
-		if err != nil {
-			return fmt.Errorf("failed to get universe domain from gcp credentials: %w", err)
```
Comment (tthvo) [inline pkg/clusterapi/system.go:319]: Maybe we don't wanna remove this code block. Let's drop the changes in this file?

**Review by tthvo — COMMENTED** [review]
/lgtm

Let's get some e2e runs to see whether this change requires any new permissions 👀
